src/main/scala/com/microsoft/partnercatalyst/fortis/spark/transforms/topic/KeywordExtractor.scala

```diff
@@ -0,0 +1,38 @@
+package com.microsoft.partnercatalyst.fortis.spark.transforms.topic
+
+import com.microsoft.partnercatalyst.fortis.spark.transforms.Tag
+import org.apache.commons.collections4.trie.PatriciaTrie
+
+import scala.collection.mutable.ListBuffer
+
+@SerialVersionUID(100L)
+class KeywordExtractor(keyWords: Seq[String]) extends Serializable {
+  private val wordTokenizer = """\b""".r
+  private val keywordTrie = new PatriciaTrie[Unit]()
+  private val originalCase = Map[String, String](keyWords.map(s => (s.toLowerCase, s)): _*)
+
+  originalCase.keys.foreach(keywordTrie.put(_, ()))
+
+  def extractKeywords(text: String): List[Tag] = {
+
+    def findMatches(segment: Seq[String]): Iterable[String] = {
```

Comment:
Note: unlike Javascript or Python, having nested functions is _not_ a performance hit in Scala. During the compile step, the inner function gets lifted to the class level so we don't re-declare it on every method call like we'd do in the dynamic languages. [Read more.](https://stackoverflow.com/a/2482410/3817588) So it's all good to use the inner functions :)